// @tanstack/start-plugin-core/dist/esm/plugin.js
import { getBundlerOptions } from "./utils.js";
import { ENTRY_POINTS, VITE_ENVIRONMENT_NAMES } from "./constants.js";
import { startManifestPlugin } from "./start-manifest-plugin/plugin.js";
import { tanStackStartRouter } from "./start-router-plugin/plugin.js";
import { loadEnvPlugin } from "./load-env-plugin/plugin.js";
import { devServerPlugin } from "./dev-server-plugin/plugin.js";
import { getClientOutputDirectory, getServerOutputDirectory } from "./output-directory.js";
import { previewServerPlugin } from "./preview-server-plugin/plugin.js";
import { parseStartConfig } from "./schema.js";
import { resolveEntry } from "./resolve-entries.js";
import { postServerBuild } from "./post-server-build.js";
import { startCompilerPlugin } from "./start-compiler-plugin/plugin.js";
import { importProtectionPlugin } from "./import-protection-plugin/plugin.js";
import { joinPaths } from "@tanstack/router-core";
import * as vite from "vite";
import { crawlFrameworkPkgs } from "vitefu";
import { join } from "pathe";
import { escapePath } from "tinyglobby";
//#region src/plugin.ts
function isFullUrl(str) {
	try {
		new URL(str);
		return true;
	} catch {
		return false;
	}
}
function TanStackStartVitePluginCore(corePluginOpts, startPluginOpts) {
	const serverFnProviderEnv = corePluginOpts.serverFn?.providerEnv || VITE_ENVIRONMENT_NAMES.server;
	const ssrIsProvider = serverFnProviderEnv === VITE_ENVIRONMENT_NAMES.server;
	const resolvedStartConfig = {
		root: "",
		startFilePath: void 0,
		routerFilePath: "",
		srcDirectory: "",
		viteAppBase: "",
		serverFnProviderEnv
	};
	let startConfig;
	const getConfig = () => {
		if (!resolvedStartConfig.root) throw new Error(`Cannot get config before root is resolved`);
		if (!startConfig) startConfig = parseStartConfig(startPluginOpts, corePluginOpts, resolvedStartConfig.root);
		return {
			startConfig,
			resolvedStartConfig,
			corePluginOpts
		};
	};
	let needsDevBaseRewrite = false;
	const capturedBundle = {};
	function getBundle(envName) {
		const bundle = capturedBundle[envName];
		if (!bundle) throw new Error(`No bundle captured for environment: ${envName}`);
		return bundle;
	}
	const environments = [{
		name: VITE_ENVIRONMENT_NAMES.client,
		type: "client"
	}, {
		name: VITE_ENVIRONMENT_NAMES.server,
		type: "server"
	}];
	if (corePluginOpts.serverFn?.providerEnv && !environments.find((e) => e.name === corePluginOpts.serverFn?.providerEnv)) environments.push({
		name: corePluginOpts.serverFn.providerEnv,
		type: "server"
	});
	return [
		{
			name: "tanstack-start-core:config",
			enforce: "pre",
			async config(viteConfig, { command }) {
				resolvedStartConfig.viteAppBase = viteConfig.base ?? "/";
				if (!isFullUrl(resolvedStartConfig.viteAppBase)) resolvedStartConfig.viteAppBase = joinPaths([
					"/",
					viteConfig.base,
					"/"
				]);
				const root = viteConfig.root || process.cwd();
				resolvedStartConfig.root = root;
				const { startConfig } = getConfig();
				if (startConfig.router.basepath === void 0) if (!isFullUrl(resolvedStartConfig.viteAppBase)) startConfig.router.basepath = resolvedStartConfig.viteAppBase.replace(/^\/|\/$/g, "");
				else startConfig.router.basepath = "/";
				else if (command === "serve" && !viteConfig.server?.middlewareMode) {
					if (!joinPaths([
						"/",
						startConfig.router.basepath,
						"/"
					]).startsWith(joinPaths([
						"/",
						resolvedStartConfig.viteAppBase,
						"/"
					]))) needsDevBaseRewrite = true;
				}
				const TSS_SERVER_FN_BASE = joinPaths([
					"/",
					startConfig.router.basepath,
					startConfig.serverFns.base,
					"/"
				]);
				const resolvedSrcDirectory = join(root, startConfig.srcDirectory);
				resolvedStartConfig.srcDirectory = resolvedSrcDirectory;
				const startFilePath = resolveEntry({
					type: "start entry",
					configuredEntry: startConfig.start.entry,
					defaultEntry: "start",
					resolvedSrcDirectory,
					required: false
				});
				resolvedStartConfig.startFilePath = startFilePath;
				const routerFilePath = resolveEntry({
					type: "router entry",
					configuredEntry: startConfig.router.entry,
					defaultEntry: "router",
					resolvedSrcDirectory,
					required: true
				});
				resolvedStartConfig.routerFilePath = routerFilePath;
				const clientEntryPath = resolveEntry({
					type: "client entry",
					configuredEntry: startConfig.client.entry,
					defaultEntry: "client",
					resolvedSrcDirectory,
					required: false
				});
				const serverEntryPath = resolveEntry({
					type: "server entry",
					configuredEntry: startConfig.server.entry,
					defaultEntry: "server",
					resolvedSrcDirectory,
					required: false
				});
				const clientAlias = vite.normalizePath(clientEntryPath ?? corePluginOpts.defaultEntryPaths.client);
				const serverAlias = vite.normalizePath(serverEntryPath ?? corePluginOpts.defaultEntryPaths.server);
				const startAlias = vite.normalizePath(startFilePath ?? corePluginOpts.defaultEntryPaths.start);
				const routerAlias = vite.normalizePath(routerFilePath);
				const entryAliasConfiguration = {
					[ENTRY_POINTS.client]: clientAlias,
					[ENTRY_POINTS.server]: serverAlias,
					[ENTRY_POINTS.start]: startAlias,
					[ENTRY_POINTS.router]: routerAlias
				};
				const startPackageName = `@tanstack/${corePluginOpts.framework}-start`;
				const crawlFrameworkPkgsResult = await crawlFrameworkPkgs({
					root: process.cwd(),
					isBuild: command === "build",
					isFrameworkPkgByJson(pkgJson) {
						const peerDependencies = pkgJson["peerDependencies"];
						if (peerDependencies) {
							if (startPackageName in peerDependencies || "@tanstack/start-client-core" in peerDependencies) return true;
						}
						return false;
					}
				});
				return {
					appType: viteConfig.appType ?? "custom",
					environments: {
						[VITE_ENVIRONMENT_NAMES.client]: {
							consumer: "client",
							build: (() => {
								const bundlerOptions = { input: { main: ENTRY_POINTS.client } };
								return {
									rollupOptions: bundlerOptions,
									rolldownOptions: bundlerOptions,
									outDir: getClientOutputDirectory(viteConfig)
								};
							})(),
							optimizeDeps: {
								exclude: crawlFrameworkPkgsResult.optimizeDeps.exclude,
								entries: [clientAlias, routerAlias].map((entry) => escapePath(entry))
							}
						},
						[VITE_ENVIRONMENT_NAMES.server]: {
							consumer: "server",
							build: {
								ssr: true,
								...(() => {
									const bundlerOptions = { input: getBundlerOptions(viteConfig.environments?.[VITE_ENVIRONMENT_NAMES.server]?.build)?.input ?? serverAlias };
									return {
										rollupOptions: bundlerOptions,
										rolldownOptions: bundlerOptions
									};
								})(),
								outDir: getServerOutputDirectory(viteConfig),
								commonjsOptions: { include: [/node_modules/] },
								copyPublicDir: viteConfig.environments?.[VITE_ENVIRONMENT_NAMES.server]?.build?.copyPublicDir ?? false
							},
							optimizeDeps: { entries: [
								serverAlias,
								startAlias,
								routerAlias
							].map((entry) => escapePath(entry)) }
						}
					},
					resolve: {
						noExternal: [
							"@tanstack/start**",
							`@tanstack/${corePluginOpts.framework}-start**`,
							...crawlFrameworkPkgsResult.ssr.noExternal.sort()
						],
						alias: { ...entryAliasConfiguration }
					},
					define: {
						...defineReplaceEnv("TSS_SERVER_FN_BASE", TSS_SERVER_FN_BASE),
						...defineReplaceEnv("TSS_ROUTER_BASEPATH", startConfig.router.basepath),
						...command === "serve" ? defineReplaceEnv("TSS_SHELL", startConfig.spa?.enabled ? "true" : "false") : {},
						...defineReplaceEnv("TSS_DEV_SERVER", command === "serve" ? "true" : "false"),
						...defineReplaceEnv("TSS_DEV_SSR_STYLES_ENABLED", startConfig.dev.ssrStyles.enabled ? "true" : "false"),
						...defineReplaceEnv("TSS_DEV_SSR_STYLES_BASEPATH", startConfig.dev.ssrStyles.basepath ?? resolvedStartConfig.viteAppBase),
						...command === "build" && startConfig.server.build.staticNodeEnv ? { "process.env.NODE_ENV": JSON.stringify(process.env.NODE_ENV || viteConfig.mode || "production") } : {}
					},
					builder: {
						sharedPlugins: true,
						async buildApp(builder) {
							const client = builder.environments[VITE_ENVIRONMENT_NAMES.client];
							const server = builder.environments[VITE_ENVIRONMENT_NAMES.server];
							if (!client) throw new Error("Client environment not found");
							if (!server) throw new Error("SSR environment not found");
							if (!client.isBuilt) await builder.build(client);
							if (!server.isBuilt) await builder.build(server);
							if (!ssrIsProvider) {
								const providerEnv = builder.environments[serverFnProviderEnv];
								if (!providerEnv) throw new Error(`Provider environment "${serverFnProviderEnv}" not found`);
								if (!providerEnv.isBuilt) await builder.build(providerEnv);
							}
						}
					}
				};
			}
		},
		{
			name: "tanstack-start-core:post-build",
			enforce: "post",
			buildApp: {
				order: "post",
				async handler(builder) {
					const { startConfig } = getConfig();
					await postServerBuild({
						builder,
						startConfig
					});
				}
			}
		},
		startCompilerPlugin({
			framework: corePluginOpts.framework,
			environments,
			generateFunctionId: startPluginOpts?.serverFns?.generateFunctionId,
			providerEnvName: serverFnProviderEnv
		}),
		importProtectionPlugin({
			getConfig,
			framework: corePluginOpts.framework,
			environments,
			providerEnvName: serverFnProviderEnv
		}),
		tanStackStartRouter(startPluginOpts, getConfig, corePluginOpts),
		loadEnvPlugin(),
		startManifestPlugin({
			getClientBundle: () => getBundle(VITE_ENVIRONMENT_NAMES.client),
			getConfig
		}),
		{
			name: "tanstack-start-core:dev-base-rewrite",
			configureServer(server) {
				if (!needsDevBaseRewrite) return;
				const basePrefix = resolvedStartConfig.viteAppBase.replace(/\/$/, "");
				server.middlewares.use((req, _res, next) => {
					if (req.url && !req.url.startsWith(basePrefix)) req.url = basePrefix + req.url;
					next();
				});
			}
		},
		devServerPlugin({
			getConfig,
			devSsrStylesEnabled: startPluginOpts?.dev?.ssrStyles?.enabled ?? true
		}),
		previewServerPlugin(),
		{
			name: "tanstack-start:core:capture-bundle",
			applyToEnvironment(e) {
				return e.name === VITE_ENVIRONMENT_NAMES.client || e.name === VITE_ENVIRONMENT_NAMES.server;
			},
			enforce: "post",
			generateBundle(_options, bundle) {
				const environment = this.environment.name;
				if (!Object.values(VITE_ENVIRONMENT_NAMES).includes(environment)) throw new Error(`Unknown environment: ${environment}`);
				capturedBundle[environment] = bundle;
			}
		}
	];
}
function defineReplaceEnv(key, value) {
	return {
		[`process.env.${key}`]: JSON.stringify(value),
		[`import.meta.env.${key}`]: JSON.stringify(value)
	};
}
//#endregion
export { TanStackStartVitePluginCore };
