// @tanstack/start-plugin-core/dist/esm/start-compiler-plugin/plugin.js
import { SERVER_FN_LOOKUP, TRANSFORM_ID_REGEX, VITE_ENVIRONMENT_NAMES } from "../constants.js";
import { cleanId } from "./utils.js";
import { KindDetectionPatterns, LookupKindsPerEnv, StartCompiler, detectKindsInCode } from "./compiler.js";
import { resolve } from "pathe";
import { VIRTUAL_MODULES } from "@tanstack/start-server-core";
import assert from "node:assert";
//#region src/start-compiler-plugin/plugin.ts
function getTransformCodeFilterForEnv(env) {
	const validKinds = LookupKindsPerEnv[env];
	const patterns = [];
	for (const [kind, pattern] of Object.entries(KindDetectionPatterns)) if (validKinds.has(kind)) patterns.push(pattern);
	return patterns;
}
var getLookupConfigurationsForEnv = (env, framework) => {
	const commonConfigs = [
		{
			libName: `@tanstack/${framework}-start`,
			rootExport: "createServerFn",
			kind: "Root"
		},
		{
			libName: `@tanstack/${framework}-start`,
			rootExport: "createIsomorphicFn",
			kind: "IsomorphicFn"
		},
		{
			libName: `@tanstack/${framework}-start`,
			rootExport: "createServerOnlyFn",
			kind: "ServerOnlyFn"
		},
		{
			libName: `@tanstack/${framework}-start`,
			rootExport: "createClientOnlyFn",
			kind: "ClientOnlyFn"
		}
	];
	if (env === "client") return [
		{
			libName: `@tanstack/${framework}-start`,
			rootExport: "createMiddleware",
			kind: "Root"
		},
		{
			libName: `@tanstack/${framework}-start`,
			rootExport: "createStart",
			kind: "Root"
		},
		...commonConfigs
	];
	else return [...commonConfigs, {
		libName: `@tanstack/${framework}-router`,
		rootExport: "ClientOnly",
		kind: "ClientOnlyJSX"
	}];
};
function resolveViteId(id) {
	return `\0${id}`;
}
var validateServerFnIdVirtualModule = `virtual:tanstack-start-validate-server-fn-id`;
function parseIdQuery(id) {
	if (!id.includes("?")) return {
		filename: id,
		query: {}
	};
	const [filename, rawQuery] = id.split(`?`, 2);
	return {
		filename,
		query: Object.fromEntries(new URLSearchParams(rawQuery))
	};
}
/**
* Generates the manifest module code for server functions.
* @param serverFnsById - Map of function IDs to their server function info
* @param includeClientReferencedCheck - Whether to include isClientReferenced flag and runtime check.
*   This is needed when SSR is NOT the provider, so server-only-referenced functions in the manifest
*   can be blocked from client HTTP requests.
*/
function generateManifestModule(serverFnsById, includeClientReferencedCheck) {
	return `
    const manifest = {${Object.entries(serverFnsById).map(([id, fn]) => {
		return `'${id}': {
                functionName: '${fn.functionName}',
        importer: () => import(${JSON.stringify(fn.extractedFilename)})${includeClientReferencedCheck ? `,
        isClientReferenced: ${fn.isClientReferenced ?? true}` : ""}
      }`;
	}).join(",")}}

    export async function getServerFnById(${includeClientReferencedCheck ? "id, opts" : "id"}) {
              const serverFnInfo = manifest[id]
              if (!serverFnInfo) {
                throw new Error('Server function info not found for ' + id)
              }
${includeClientReferencedCheck ? `
      // If called from client, only allow client-referenced functions
      if (opts?.fromClient && !serverFnInfo.isClientReferenced) {
        throw new Error('Server function not accessible from client: ' + id)
      }
` : ""}
              const fnModule = await serverFnInfo.importer()

              if (!fnModule) {
                console.info('serverFnInfo', serverFnInfo)
                throw new Error('Server function module not resolved for ' + id)
              }

              const action = fnModule[serverFnInfo.functionName]

              if (!action) {
                  console.info('serverFnInfo', serverFnInfo)
                  console.info('fnModule', fnModule)

                throw new Error(
                  \`Server function module export not resolved for serverFn ID: \${id}\`,
                )
              }
              return action
            }
          `;
}
function startCompilerPlugin(opts) {
	const compilers = {};
	const serverFnsById = {};
	const onServerFnsById = (d) => {
		Object.assign(serverFnsById, d);
	};
	let root = process.cwd();
	const resolvedResolverVirtualImportId = resolveViteId(VIRTUAL_MODULES.serverFnResolver);
	const ssrEnvName = VITE_ENVIRONMENT_NAMES.server;
	const ssrIsProvider = opts.providerEnvName === ssrEnvName;
	const appliedResolverEnvironments = new Set(ssrIsProvider ? [opts.providerEnvName] : [ssrEnvName, opts.providerEnvName]);
	function perEnvServerFnPlugin(environment) {
		const transformCodeFilter = getTransformCodeFilterForEnv(environment.type);
		return {
			name: `tanstack-start-core::server-fn:${environment.name}`,
			enforce: "pre",
			applyToEnvironment(env) {
				return env.name === environment.name;
			},
			configResolved(config) {
				root = config.root;
				config.command;
			},
			transform: {
				filter: {
					id: {
						exclude: new RegExp(`${SERVER_FN_LOOKUP}$`),
						include: TRANSFORM_ID_REGEX
					},
					code: { include: transformCodeFilter }
				},
				async handler(code, id) {
					let compiler = compilers[this.environment.name];
					if (!compiler) {
						const mode = this.environment.mode === "build" ? "build" : "dev";
						compiler = new StartCompiler({
							env: environment.type,
							envName: environment.name,
							root,
							lookupKinds: LookupKindsPerEnv[environment.type],
							lookupConfigurations: getLookupConfigurationsForEnv(environment.type, opts.framework),
							mode,
							framework: opts.framework,
							providerEnvName: opts.providerEnvName,
							generateFunctionId: opts.generateFunctionId,
							onServerFnsById,
							getKnownServerFns: () => serverFnsById,
							loadModule: async (id) => {
								if (this.environment.mode === "build") {
									const code = (await this.load({ id })).code ?? "";
									compiler.ingestModule({
										code,
										id
									});
								} else if (this.environment.mode === "dev")
 /**
								* in dev, vite does not return code from `ctx.load()`
								* so instead, we need to take a different approach
								* we must force vite to load the module and run it through the vite plugin pipeline
								* we can do this by using the `fetchModule` method
								* the `captureServerFnModuleLookupPlugin` captures the module code via its transform hook and invokes analyzeModuleAST
								*/
								await this.environment.fetchModule(id + "?" + SERVER_FN_LOOKUP);
								else throw new Error(`could not load module ${id}: unknown environment mode ${this.environment.mode}`);
							},
							resolveId: async (source, importer) => {
								const r = await this.resolve(source, importer);
								if (r) {
									if (!r.external) return cleanId(r.id);
								}
								return null;
							}
						});
						compilers[this.environment.name] = compiler;
					}
					const detectedKinds = detectKindsInCode(code, environment.type);
					return await compiler.compile({
						id,
						code,
						detectedKinds
					});
				}
			},
			hotUpdate(ctx) {
				const compiler = compilers[this.environment.name];
				ctx.modules.forEach((m) => {
					if (m.id) {
						if (compiler?.invalidateModule(m.id)) m.importers.forEach((importer) => {
							if (importer.id) compiler?.invalidateModule(importer.id);
						});
					}
				});
			}
		};
	}
	return [
		...opts.environments.map(perEnvServerFnPlugin),
		{
			name: "tanstack-start-core:capture-server-fn-module-lookup",
			apply: "serve",
			applyToEnvironment(env) {
				return !!opts.environments.find((e) => e.name === env.name);
			},
			transform: {
				filter: { id: new RegExp(`${SERVER_FN_LOOKUP}$`) },
				handler(code, id) {
					compilers[this.environment.name]?.ingestModule({
						code,
						id: cleanId(id)
					});
				}
			}
		},
		{
			name: "tanstack-start-core:validate-server-fn-id",
			apply: "serve",
			load: {
				filter: { id: new RegExp(resolveViteId(validateServerFnIdVirtualModule)) },
				async handler(id) {
					const fnId = parseIdQuery(id).query.id;
					if (fnId && serverFnsById[fnId]) return `export {}`;
					if (fnId) try {
						const decoded = JSON.parse(Buffer.from(fnId, "base64url").toString("utf8"));
						if (typeof decoded.file === "string" && typeof decoded.export === "string") {
							let sourceFile = decoded.file;
							if (sourceFile.startsWith("/@id/")) sourceFile = sourceFile.slice(5);
							const qIdx = sourceFile.indexOf("?");
							if (qIdx !== -1) sourceFile = sourceFile.slice(0, qIdx);
							const absPath = resolve(root, sourceFile);
							assert(this.environment.mode === "dev");
							await this.environment.fetchModule(absPath);
							if (serverFnsById[fnId]) return `export {}`;
						}
					} catch {}
					this.error(`Invalid server function ID: ${fnId}`);
				}
			}
		},
		{
			name: "tanstack-start-core:server-fn-resolver",
			enforce: "pre",
			applyToEnvironment: (env) => {
				return appliedResolverEnvironments.has(env.name);
			},
			configResolved(config) {
				root = config.root;
				config.command;
			},
			resolveId: {
				filter: { id: new RegExp(VIRTUAL_MODULES.serverFnResolver) },
				handler() {
					return resolvedResolverVirtualImportId;
				}
			},
			load: {
				filter: { id: new RegExp(resolvedResolverVirtualImportId) },
				handler() {
					if (this.environment.name !== opts.providerEnvName) return `export { getServerFnById } from '@tanstack/start-server-core/server-fn-ssr-caller'`;
					if (this.environment.mode !== "build") return `
            export async function getServerFnById(id) {
              const validateIdImport = ${JSON.stringify(validateServerFnIdVirtualModule)} + '?id=' + id
              await import(/* @vite-ignore */ '/@id/__x00__' + validateIdImport)
              const decoded = Buffer.from(id, 'base64url').toString('utf8')
              const devServerFn = JSON.parse(decoded)
              const mod = await import(/* @vite-ignore */ devServerFn.file)
              return mod[devServerFn.export]
            }
            `;
					return generateManifestModule(serverFnsById, !ssrIsProvider);
				}
			}
		}
	];
}
//#endregion
export { startCompilerPlugin };
