// @tanstack/start-plugin-core/dist/esm/start-manifest-plugin/manifestBuilder.js
import { resolveManifestAssetLink, rootRouteId } from "@tanstack/router-core";
import { joinURL } from "ufo";
import { tsrSplit } from "@tanstack/router-plugin";
//#region src/start-manifest-plugin/manifestBuilder.ts
var ROUTER_MANAGED_MODE = 1;
var NON_ROUTE_DYNAMIC_MODE = 2;
var VISITING_CHUNK = 1;
function appendUniqueStrings(target, source) {
	if (source.length === 0) return target;
	if (!target || target.length === 0) return source;
	const seen = new Set(target);
	let result;
	for (const value of source) {
		if (seen.has(value)) continue;
		seen.add(value);
		if (!result) result = target.slice();
		result.push(value);
	}
	return result ?? target;
}
function appendUniqueAssets(target, source) {
	if (source.length === 0) return target;
	if (!target || target.length === 0) return source;
	const seen = new Set(target.map(getAssetIdentity));
	let result;
	for (const asset of source) {
		const identity = getAssetIdentity(asset);
		if (seen.has(identity)) continue;
		seen.add(identity);
		if (!result) result = target.slice();
		result.push(asset);
	}
	return result ?? target;
}
function getAssetIdentity(asset) {
	if (asset.tag === "link" || asset.tag === "script") {
		const attrs = asset.attrs ?? {};
		return [
			asset.tag,
			"href" in attrs ? String(attrs.href) : "",
			"src" in attrs ? String(attrs.src) : "",
			"rel" in attrs ? String(attrs.rel) : "",
			"type" in attrs ? String(attrs.type) : "",
			asset.children ?? ""
		].join("|");
	}
	return JSON.stringify(asset);
}
function mergeRouteChunkData(options) {
	const chunkAssets = options.getChunkCssAssets(options.chunk);
	const chunkPreloads = options.getChunkPreloads(options.chunk);
	options.route.assets = appendUniqueAssets(options.route.assets, chunkAssets);
	options.route.preloads = appendUniqueStrings(options.route.preloads, chunkPreloads);
}
function buildStartManifest(options) {
	const scannedChunks = scanClientChunks(options.clientBundle);
	const hashedCssFiles = collectDynamicImportCss(scannedChunks.routeEntryChunks, scannedChunks.chunksByFileName, scannedChunks.entryChunk);
	const assetResolvers = createManifestAssetResolvers({
		basePath: options.basePath,
		hashedCssFiles
	});
	const routes = buildRouteManifestRoutes({
		routeTreeRoutes: options.routeTreeRoutes,
		routeChunksByFilePath: scannedChunks.routeChunksByFilePath,
		chunksByFileName: scannedChunks.chunksByFileName,
		entryChunk: scannedChunks.entryChunk,
		assetResolvers
	});
	dedupeNestedRouteManifestEntries(rootRouteId, routes[rootRouteId], routes);
	for (const routeId of Object.keys(routes)) {
		const route = routes[routeId];
		const hasAssets = route.assets && route.assets.length > 0;
		const hasPreloads = route.preloads && route.preloads.length > 0;
		if (!hasAssets && !hasPreloads) delete routes[routeId];
	}
	return {
		routes,
		clientEntry: assetResolvers.getAssetPath(scannedChunks.entryChunk.fileName)
	};
}
function scanClientChunks(clientBundle) {
	let entryChunk;
	const chunksByFileName = /* @__PURE__ */ new Map();
	const routeChunksByFilePath = /* @__PURE__ */ new Map();
	const routeEntryChunks = /* @__PURE__ */ new Set();
	for (const fileName in clientBundle) {
		const bundleEntry = clientBundle[fileName];
		if (bundleEntry.type !== "chunk") continue;
		chunksByFileName.set(bundleEntry.fileName, bundleEntry);
		if (bundleEntry.isEntry) {
			if (entryChunk) throw new Error(`multiple entries detected: ${entryChunk.fileName} ${bundleEntry.fileName}`);
			entryChunk = bundleEntry;
		}
		const routeFilePaths = getRouteFilePathsFromModuleIds(bundleEntry.moduleIds);
		if (routeFilePaths.length === 0) continue;
		routeEntryChunks.add(bundleEntry);
		for (let i = 0; i < routeFilePaths.length; i++) {
			const routeFilePath = routeFilePaths[i];
			let chunks = routeChunksByFilePath.get(routeFilePath);
			if (chunks === void 0) {
				chunks = [];
				routeChunksByFilePath.set(routeFilePath, chunks);
			}
			chunks.push(bundleEntry);
		}
	}
	if (!entryChunk) throw new Error("No entry file found");
	return {
		entryChunk,
		chunksByFileName,
		routeChunksByFilePath,
		routeEntryChunks
	};
}
function getRouteFilePathsFromModuleIds(moduleIds) {
	let routeFilePaths;
	let seenRouteFilePaths;
	for (const moduleId of moduleIds) {
		const queryIndex = moduleId.indexOf("?");
		if (queryIndex < 0) continue;
		const query = moduleId.slice(queryIndex + 1);
		if (!query.includes(tsrSplit)) continue;
		if (!new URLSearchParams(query).has(tsrSplit)) continue;
		const routeFilePath = moduleId.slice(0, queryIndex);
		if (seenRouteFilePaths?.has(routeFilePath)) continue;
		if (routeFilePaths === void 0) {
			routeFilePaths = [];
			seenRouteFilePaths = /* @__PURE__ */ new Set();
		}
		routeFilePaths.push(routeFilePath);
		seenRouteFilePaths.add(routeFilePath);
	}
	return routeFilePaths ?? [];
}
function collectDynamicImportCss(routeEntryChunks, chunksByFileName, entryChunk) {
	const routerManagedCssFiles = /* @__PURE__ */ new Set();
	const nonRouteDynamicCssFiles = /* @__PURE__ */ new Set();
	const hashedCssFiles = /* @__PURE__ */ new Set();
	const visitedByChunk = /* @__PURE__ */ new Map();
	const chunkStack = [];
	const modeStack = [];
	for (const routeEntryChunk of routeEntryChunks) {
		chunkStack.push(routeEntryChunk);
		modeStack.push(ROUTER_MANAGED_MODE);
	}
	if (entryChunk) {
		chunkStack.push(entryChunk);
		modeStack.push(ROUTER_MANAGED_MODE);
	}
	while (chunkStack.length > 0) {
		const chunk = chunkStack.pop();
		const mode = modeStack.pop();
		const previousMode = visitedByChunk.get(chunk) ?? 0;
		if ((previousMode & mode) === mode) continue;
		visitedByChunk.set(chunk, previousMode | mode);
		if ((mode & ROUTER_MANAGED_MODE) !== 0) for (const cssFile of chunk.viteMetadata?.importedCss ?? []) routerManagedCssFiles.add(cssFile);
		if ((mode & NON_ROUTE_DYNAMIC_MODE) !== 0) for (const cssFile of chunk.viteMetadata?.importedCss ?? []) nonRouteDynamicCssFiles.add(cssFile);
		for (let i = 0; i < chunk.imports.length; i++) {
			const importedChunk = chunksByFileName.get(chunk.imports[i]);
			if (importedChunk) {
				chunkStack.push(importedChunk);
				modeStack.push(mode);
			}
		}
		for (let i = 0; i < chunk.dynamicImports.length; i++) {
			const dynamicImportedChunk = chunksByFileName.get(chunk.dynamicImports[i]);
			if (dynamicImportedChunk) {
				chunkStack.push(dynamicImportedChunk);
				modeStack.push((mode & NON_ROUTE_DYNAMIC_MODE) !== 0 || !routeEntryChunks.has(dynamicImportedChunk) ? NON_ROUTE_DYNAMIC_MODE : ROUTER_MANAGED_MODE);
			}
		}
	}
	for (const cssFile of routerManagedCssFiles) if (nonRouteDynamicCssFiles.has(cssFile)) hashedCssFiles.add(cssFile);
	return hashedCssFiles;
}
function createManifestAssetResolvers(options) {
	const assetPathByFileName = /* @__PURE__ */ new Map();
	const stylesheetAssetByFileName = /* @__PURE__ */ new Map();
	const preloadsByChunk = /* @__PURE__ */ new Map();
	const getAssetPath = (fileName) => {
		const cachedPath = assetPathByFileName.get(fileName);
		if (cachedPath) return cachedPath;
		const assetPath = joinURL(options.basePath, fileName);
		assetPathByFileName.set(fileName, assetPath);
		return assetPath;
	};
	const getStylesheetAsset = (cssFile) => {
		const cachedAsset = stylesheetAssetByFileName.get(cssFile);
		if (cachedAsset) return cachedAsset;
		const href = getAssetPath(cssFile);
		const asset = {
			tag: "link",
			attrs: {
				rel: "stylesheet",
				href: options.hashedCssFiles?.has(cssFile) ? `${href}#` : href,
				type: "text/css"
			}
		};
		stylesheetAssetByFileName.set(cssFile, asset);
		return asset;
	};
	const getChunkPreloads = (chunk) => {
		const cachedPreloads = preloadsByChunk.get(chunk);
		if (cachedPreloads) return cachedPreloads;
		const preloads = [getAssetPath(chunk.fileName)];
		for (let i = 0; i < chunk.imports.length; i++) preloads.push(getAssetPath(chunk.imports[i]));
		preloadsByChunk.set(chunk, preloads);
		return preloads;
	};
	return {
		getAssetPath,
		getChunkPreloads,
		getStylesheetAsset
	};
}
function createChunkCssAssetCollector(options) {
	const assetsByChunk = /* @__PURE__ */ new Map();
	const stateByChunk = /* @__PURE__ */ new Map();
	const appendAsset = (assets, seenAssets, asset) => {
		const identity = getAssetIdentity(asset);
		if (seenAssets.has(identity)) return;
		seenAssets.add(identity);
		assets.push(asset);
	};
	const getChunkCssAssets = (chunk) => {
		const cachedAssets = assetsByChunk.get(chunk);
		if (cachedAssets) return cachedAssets;
		if (stateByChunk.get(chunk) === VISITING_CHUNK) return [];
		stateByChunk.set(chunk, VISITING_CHUNK);
		const assets = [];
		const seenAssets = /* @__PURE__ */ new Set();
		for (const cssFile of chunk.viteMetadata?.importedCss ?? []) appendAsset(assets, seenAssets, options.getStylesheetAsset(cssFile));
		for (let i = 0; i < chunk.imports.length; i++) {
			const importedChunk = options.chunksByFileName.get(chunk.imports[i]);
			if (!importedChunk) continue;
			const importedAssets = getChunkCssAssets(importedChunk);
			for (let j = 0; j < importedAssets.length; j++) appendAsset(assets, seenAssets, importedAssets[j]);
		}
		stateByChunk.delete(chunk);
		assetsByChunk.set(chunk, assets);
		return assets;
	};
	return { getChunkCssAssets };
}
function buildRouteManifestRoutes(options) {
	const routes = {};
	const getChunkCssAssets = createChunkCssAssetCollector({
		chunksByFileName: options.chunksByFileName,
		getStylesheetAsset: options.assetResolvers.getStylesheetAsset
	}).getChunkCssAssets;
	for (const [routeId, route] of Object.entries(options.routeTreeRoutes)) {
		if (!route.filePath) {
			if (routeId === rootRouteId) {
				routes[routeId] = route;
				continue;
			}
			throw new Error(`expected filePath to be set for ${routeId}`);
		}
		const chunks = options.routeChunksByFilePath.get(route.filePath);
		if (!chunks) {
			routes[routeId] = route;
			continue;
		}
		const existing = routes[routeId];
		const targetRoute = routes[routeId] = existing ? existing : { ...route };
		for (const chunk of chunks) mergeRouteChunkData({
			route: targetRoute,
			chunk,
			getChunkCssAssets,
			getChunkPreloads: options.assetResolvers.getChunkPreloads
		});
	}
	mergeRouteChunkData({
		route: routes[rootRouteId] = routes[rootRouteId] || {},
		chunk: options.entryChunk,
		getChunkCssAssets,
		getChunkPreloads: options.assetResolvers.getChunkPreloads
	});
	return routes;
}
function dedupeNestedRouteManifestEntries(routeId, route, routesById, seenPreloads = /* @__PURE__ */ new Set(), seenAssets = /* @__PURE__ */ new Set()) {
	let routePreloads = route.preloads;
	let routeAssets = route.assets;
	if (routePreloads && routePreloads.length > 0) {
		let dedupedPreloads;
		for (let i = 0; i < routePreloads.length; i++) {
			const preload = routePreloads[i];
			const preloadHref = resolveManifestAssetLink(preload).href;
			if (seenPreloads.has(preloadHref)) {
				if (dedupedPreloads === void 0) dedupedPreloads = routePreloads.slice(0, i);
				continue;
			}
			seenPreloads.add(preloadHref);
			if (dedupedPreloads) dedupedPreloads.push(preload);
		}
		if (dedupedPreloads) {
			routePreloads = dedupedPreloads;
			route.preloads = dedupedPreloads;
		}
	}
	if (routeAssets && routeAssets.length > 0) {
		let dedupedAssets;
		for (let i = 0; i < routeAssets.length; i++) {
			const asset = routeAssets[i];
			const identity = getAssetIdentity(asset);
			if (seenAssets.has(identity)) {
				if (dedupedAssets === void 0) dedupedAssets = routeAssets.slice(0, i);
				continue;
			}
			seenAssets.add(identity);
			if (dedupedAssets) dedupedAssets.push(asset);
		}
		if (dedupedAssets) {
			routeAssets = dedupedAssets;
			route.assets = dedupedAssets;
		}
	}
	if (route.children) for (const childRouteId of route.children) {
		const childRoute = routesById[childRouteId];
		if (!childRoute) throw new Error(`Route tree references child route ${childRouteId} from ${routeId}, but no route entry was found`);
		dedupeNestedRouteManifestEntries(childRouteId, childRoute, routesById, seenPreloads, seenAssets);
	}
	if (routePreloads) for (let i = routePreloads.length - 1; i >= 0; i--) seenPreloads.delete(resolveManifestAssetLink(routePreloads[i]).href);
	if (routeAssets) for (let i = routeAssets.length - 1; i >= 0; i--) seenAssets.delete(getAssetIdentity(routeAssets[i]));
}
//#endregion
export { buildStartManifest };
