// @tanstack/start-plugin-core/dist/esm/start-manifest-plugin/plugin.js
import { resolveViteId } from "../utils.js";
import { ENTRY_POINTS } from "../constants.js";
import { buildStartManifest } from "./manifestBuilder.js";
import { joinURL } from "ufo";
import { VIRTUAL_MODULES } from "@tanstack/start-server-core";
//#region src/start-manifest-plugin/plugin.ts
var resolvedModuleId = resolveViteId(VIRTUAL_MODULES.startManifest);
function startManifestPlugin(opts) {
	return {
		name: "tanstack-start:start-manifest-plugin",
		enforce: "pre",
		resolveId: {
			filter: { id: new RegExp(VIRTUAL_MODULES.startManifest) },
			handler(id) {
				if (id === VIRTUAL_MODULES.startManifest) return resolvedModuleId;
			}
		},
		load: {
			filter: { id: new RegExp(resolvedModuleId) },
			handler(id) {
				const { resolvedStartConfig } = opts.getConfig();
				if (id === resolvedModuleId) {
					if (this.environment.name !== resolvedStartConfig.serverFnProviderEnv) return `export default {}`;
					if (this.environment.config.command === "serve") return `export const tsrStartManifest = () => ({
            routes: {},
            clientEntry: '${joinURL(resolvedStartConfig.viteAppBase, "@id", ENTRY_POINTS.client)}',
          })`;
					const routeTreeRoutes = globalThis.TSS_ROUTES_MANIFEST;
					const startManifest = buildStartManifest({
						clientBundle: opts.getClientBundle(),
						routeTreeRoutes,
						basePath: resolvedStartConfig.viteAppBase
					});
					return `export const tsrStartManifest = () => (${JSON.stringify(startManifest)})`;
				}
			}
		}
	};
}
//#endregion
export { startManifestPlugin };
